### ISSUE TITLE
`SMTP` - Email server uptime checking monitor

### ISSUE THREAD

**Issue by ChrisThePCGeek** [louislam/uptime-kuma#3023]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Monitor

### 🔖 Feature description

- send an email using a specified smtp server
- said server either itself or some service perhaps (not sure if one exists) that would then report back "hey i got the email" to verify it is up

### ✔️ Solution

noted in description

**Comment by stefanux** [comment]
so you basically want a email-sent and receive loop and have a notification if no mail show up in interval x, right?

**Comment by ChrisThePCGeek** [comment]
Yeah that's the idea I had, I think it would work to verify an email server was up and working.  could even trigger a fail if the send doesn't work since sometimes send works but receive doesn't.

**Comment by solracsf** [comment]
Something similar to these options would be great.
https://www.wormly.com/test-smtp-server

**Comment by meyerf99** [comment]
@ChrisThePCGeek Interested to develop together a new monitor for mail server's?

**Comment by ediazrod** [comment]
I love to have this feature, at least to test if the e-mail works and on the other end if the mail is received... I love to follow up this capability

**Comment by erseco** [comment]
+1

**Comment by CommanderStorm** [comment]
@ediazrod @erseco 

Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder.
Issues are for discussing **what needs to be done how by whom**.
[We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

**Comment by hujiko** [comment]
I think the sending and receiving actual Mail might be quite complex.

I would already be fine with something more simple. A check, that
- Connects to a Server/Port
- Issues StartTLS if checked in the check options
- Validates the SSL-Certificate (either when StartTLS is configured, or as separate switch)
- Sends "MAIL FROM: .." with a from-address defined in the check and expects a 250 response
- Sends "RCPT TO: ..." with a to-address defined in the check, and expects a 250 response.
And then the check can terminate the connection. It does not actually send the mail, but goes far enough to check the SSL config, the (potential) config for sender and receive adresses.

Something like:

```javascript
const { SMTPClient } = require('smtp-client');

module.exports = function(smtpConfig) {
  return {
    async check() {
      const client = new SMTPClient({
        host: smtpConfig.host,
        port: smtpConfig.port,
        secure: false,
        useSecureTransport: false,
        ignoreTLS: true,
        requireTLS: false,
        timeout: 5000,
      });

      try {
        await client.connect();

        const response = await client.sendCommand('EHLO localhost');
        if (response.code !== 250) {
          throw new Error('EHLO command failed');
        }

        const starttlsResponse = await client.sendCommand('STARTTLS');
        if (starttlsResponse.code !== 220) {
          throw new Error('STARTTLS command failed');
        }

        await client.secure({
          rejectUnauthorized: smtpConfig.rejectUnauthorized,
        });

        const mailFromResponse = await client.sendCommand(`MAIL FROM:<${smtpConfig.from}>`);
        if (mailFromResponse.code !== 250) {
          throw new Error('MAIL FROM command failed');
        }

        const rcptToResponse = await client.sendCommand(`RCPT TO:<${smtpConfig.to}>`);
        if (rcptToResponse.code !== 250) {
          throw new Error('RCPT TO command failed');
        }

        return true; // SMTP server is reachable
      } catch (error) {
        throw new Error(`SMTP server check failed: ${error.message}`);
      } finally {
        client.close();
      }
    },
  };
};
```

Whats missing there is the ability to make StartTLS optional and to actually validate the Certificate.

**Comment by CommanderStorm** [comment]
Note that our contribution guide is here https://github.com/louislam/uptime-kuma/blob/1550a5f79270f39007086d8d23c180ba9f63096b/CONTRIBUTING.md ^^

**Comment by mkarg** [comment]
I would really love to see support for uptime checking of SMTP MTAs in Uptime-Kuma.

In fact I think it would be enough to be able to connect and receive an initial positive status (i. e. the full roundtrip should be *optional* to configure).

**Comment by arturhefczyc** [comment]
I am willing to sponsor implementation of this feature. If someone is interested in doing this please let me know.

**Comment by hadifarnoud** [comment]
any update on this? seems super useful

**Comment by tvdsluijs** [comment]
Great suggestion!

**Comment by apio-sys** [comment]
I think the scope of Uptime Kuma is to measure if something is indeed UP. I monitor my email servers by checking that port 25 responds. If it does, that means my MTA - usually postfix - is actually listening and working. However, this would not give you the health of all the underlying services (usually quite a lot for an email server). And IMHO that is beyond an uptime checking service. And sending emails back and forth is a bad idea that might cause backscatter, blocks by postscreen or fail2ban and other NDR's.

**Comment by hadifarnoud** [comment]
having a reponse check for a port is going a long way for any sysadmin. sending emails is complicated to check I know.

I would be happy knowing at least the server is up

**Comment by koehn** [comment]
FWIW in #5489 (included in 2.0.0-beta.3) is an SMTP monitor that confirms that the server is actually functioning. This is roughly analogous to the HTTP checks in Uptime-Kuma.

**Comment by CommanderStorm** [comment]
Thanks for the heads up, we forgot to close this one I guess.

Yes, the FR was "send an email", but I doubt that people actually want this. That is how you get on spam lists fast.
